### PR TITLE
new implementation of build phase commands

### PR DESCRIPTION
**Pull request by dhellmann**
Separate phases for downloading source archive, unpacking and patching it,
preparing a build environment, then building a wheel. Includes a test
script to demonstrate how the phases pass data from one to the next,
but we can almost certainly improve on that using a more structured
state file.

### CODE REVIEW

**Comment by dhellmann** [comment]
This should replace #54 and #57, but I will leave them open for reference for now.

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -59,3 +59,35 @@ jobs:
...
+
+      - name: Setup Podman
+        run: |
+          sudo apt update
+          sudo apt-get -y install podman
+          podman pull fedora:39
+
+      - name: Run tests
+        run: |
+          ${{ matrix.test-script }}
```
Comment (markmc) [inline .github/workflows/test.yaml:93]: Missing `PYTHON=python${{ matrix.python-version }}` I guess

File: e2e/test_rebuild_following_bootstrap.sh [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+VENV=$WORKDIR/venv-mirror-tools
+python3 -m venv $VENV
+$VENV/bin/python3 -m pip install python-pypi-mirror
+
+build_wheel() {
+    local name="$1"; shift
+    local version="$1"; shift
+
+    in_container download-source-archive "$name" "$version" 2>&1 | tee $WORKDIR/download-source-archive-$name.log
+    source_filename=$(cat $WORKDIR/last-download.txt)
```
Comment (markmc) [inline e2e/test_rebuild_following_bootstrap.sh:95]: Wouldn't be a huge fan of this pattern of returning a filename by storing it in a file, but I guess you're not either and have tried other things :)

File: mirror_builder/__main__.py [hunk 3]
```diff
@@ -51,5 +75,41 @@ def do_bootstrap(ctx, args):
...
+    req = Requirement(f'{args.dist_name}=={args.dist_version}')
+    source_root_dir = pathlib.Path(args.source_dir)
+    sdist.prepare_build_environment(ctx, req, source_root_dir)
+
+
+def do_build(ctx, args):
+    req = Requirement(f'{args.dist_name}=={args.dist_version}')
+    source_root_dir = pathlib.Path(args.source_dir)
+    build_env = wheels.BuildEnvironment(ctx, source_root_dir.parent, None)
+    wheel_filenames = wheels.build_wheel(ctx, req, source_root_dir, build_env)
```
Comment (markmc) [inline mirror_builder/__main__.py:107]: I know it's pre-existing, but this thing about "user: build me a wheel", "program: here's a list of wheels that we expect to only have one element" is pretty annoying

File: mirror_builder/wheels.py [hunk 4]
```diff
@@ -42,18 +37,22 @@ class BuildEnvironment:
...
         self._createenv()
 
     @property
     def python(self):
-        return (self._path / 'bin/python3').absolute()
+        return (self.path / 'bin/python3').absolute()
 
     def _createenv(self):
+        if self.path.exists():
+            logger.info('reusing build environment in %s', self.path)
```
Comment (markmc) [inline mirror_builder/wheels.py:50]: This made a bit nervous at first ... might we end up reusing a build environment between builds, inadvertently?

I think I'd like some safety here - like some way of knowing this is a just-prepared-for-this-exact-build environment that we're "reusing"

**Review by markmc — COMMENTED** [review]
Looks excellent, thanks Doug

**Comment by markmc** [comment]
Merging, even there's a few things I commented on that might be worth following up on

Ignoring the CI failure because I think it's just the `PYTHON` thing - it worked for 3.12

File: .github/workflows/test.yaml (hunk 1 above)
Comment (markmc) [reply .github/workflows/test.yaml:93]: Fixed by #65

File: mirror_builder/__main__.py (hunk 3 above)
Comment (markmc) [reply mirror_builder/__main__.py:107]: Filed as #66

File: mirror_builder/wheels.py (hunk 4 above)
Comment (markmc) [reply mirror_builder/wheels.py:50]: Filed as #67

**Pull request merged**